backend/msgpack/placement.go

```diff
@@ -0,0 +1,269 @@
+	NumShards() (int, error)
+
+	// Shard returns the shard associated with a given metric id.
+	Shard(id []byte) (uint32, error)
+
+	// RouteShard routes msgpack encoded buffer for a given shard.
+	RouteShard(shard uint32, buf msgpack.Buffer) error
+
+	// Close closes the topology.
+	Close() error
+}
+
+type activePlacement struct {
+	sync.RWMutex
+
+	snapshots []*placementSnapshot
+	nowFn     clock.NowFn
+	queueMgr  instanceQueueManger
+	expiring  int32
+	closed    bool
+}
+
+func newActivePlacement(snapshots []*placementSnapshot, opts placementOptions) placement {
+	queueMgr := opts.InstanceQueueManger()
+	for _, snapshot := range snapshots {
+		queueMgr.AddInstances(snapshot.Instances())
+	}
+	return &activePlacement{
+		snapshots: snapshots,
+		nowFn:     opts.ClockOptions().NowFn(),
+		queueMgr:  queueMgr,
+	}
+}
+
+func (p *activePlacement) NumShards() (int, error) {
+	p.RLock()
+	snapshot, err := p.activeSnapshotWithLock(p.nowFn())
+	if err != nil {
+		p.RUnlock()
+		return 0, err
+	}
+	numShards := snapshot.NumShards()
+	p.RUnlock()
+	return numShards, nil
+}
+
+func (p *activePlacement) Shard(id []byte) (uint32, error) {
+	p.RLock()
+	snapshot, err := p.activeSnapshotWithLock(p.nowFn())
+	if err != nil {
+		p.RUnlock()
+		return 0, err
+	}
+	shard := snapshot.Shard(id)
+	p.RUnlock()
+	return shard, nil
+}
+
+func (p *activePlacement) RouteShard(shard uint32, buf msgpack.Buffer) error {
+	p.RLock()
+	snapshot, err := p.activeSnapshotWithLock(p.nowFn())
+	if err != nil {
+		p.RUnlock()
+		return err
+	}
+	instances := snapshot.InstancesForShard(shard)
+	err = p.queueMgr.WriteTo(instances, buf)
+	p.RUnlock()
+	return err
+}
+
+func (p *activePlacement) Close() error {
+	p.Lock()
+	defer p.Unlock()
+
+	if p.closed {
+		return errPlacementClosed
+	}
+	for _, snapshot := range p.snapshots {
+		p.queueMgr.RemoveInstances(snapshot.Instances())
+	}
+	p.snapshots = nil
+	return nil
+}
+
+func (p *activePlacement) activeSnapshotWithLock(t time.Time) (*placementSnapshot, error) {
+	if p.closed {
+		return nil, errPlacementClosed
+	}
+	idx := p.activeIndexWithLock(t)
+	if idx < 0 {
+		return nil, errNoApplicablePlacementSnapshot
+	}
+	snapshot := p.snapshots[idx]
+	// If the snapshot that's in effect is not the first placment, expire the stale ones.
+	if idx > 0 && atomic.CompareAndSwapInt32(&p.expiring, 0, 1) {
+		go p.expire()
+	}
+	return snapshot, nil
+}
+
+func (p *activePlacement) activeIndexWithLock(t time.Time) int {
+	timeNs := t.UnixNano()
+	idx := len(p.snapshots) - 1
+	for idx >= 0 && p.snapshots[idx].cutoverNs > timeNs {
+		idx--
+	}
+	return idx
+}
+
+func (p *activePlacement) expire() {
+	// NB(xichen): this improves readability at the slight cost of lambda capture
+	// because this code path is triggered very infrequently.
+	cleanup := func() {
+		p.Unlock()
+		atomic.StoreInt32(&p.expiring, 0)
+	}
+	p.Lock()
+	defer cleanup()
+
+	if p.closed {
+		return
+	}
+	idx := p.activeIndexWithLock(p.nowFn())
```

Comment:
Are these read only operations? If so, would it be more efficient to have this and line 165 inside an RLock?